ompi/runtime/ompi_spc.c

```diff
@@ -0,0 +1,510 @@
+ * to the correct value for this pvar.
+ */
+static int ompi_spc_get_count(const struct mca_base_pvar_t *pvar, void *value, void *obj_handle)
+{   
+    (void) obj_handle;
+
+    long long *counter_value = (long long*)value;
+
+    if(OPAL_UNLIKELY(mpi_t_disabled == 1)) {
+        *counter_value = 0;
+        return MPI_SUCCESS;
+    }
+
+    /* Convert from MPI_T pvar index to SPC index */
+    int index = pvar->pvar_index - mpi_t_offset;
+    /* Set the counter value to the current SPC value */
+    *counter_value = events[index].value;
+    /* If this is a timer-based counter, convert from cycles to microseconds */
+    if(timer_event[index])
+        *counter_value /= sys_clock_freq_mhz;
+    /* If this is a high watermark counter, reset it after it has been read */
+    if(index == OMPI_MAX_UNEXPECTED_IN_QUEUE || index == OMPI_MAX_OOS_IN_QUEUE)
+        events[index].value = 0;
+
+    return MPI_SUCCESS;
+}
+
+/* Initializes the events data structure and allocates memory for it if needed. */
+void events_init()
+{
+    int i;
+
+    /* If the events data structure hasn't been allocated yet, allocate memory for it */
+    if(events == NULL) {
+        events = (ompi_event_t*)malloc(OMPI_NUM_COUNTERS * sizeof(ompi_event_t));
+    }
+    /* The data structure has been allocated, so we simply initialize all of the counters
+     * with their names and an initial count of 0.
+     */
+    for(i = 0; i < OMPI_NUM_COUNTERS; i++) {
+        events[i].name = counter_names[i];
+        events[i].value = 0;
+    }
+}
+
+/* Initializes the SPC data structures and registers all counters as MPI_T pvars.
+ * Turns on only the counters that were specified in the mpi_spc_attach MCA parameter.  
+ */
+void ompi_spc_init()
+{
+    int i, j, ret, found = 0, all_on = 0;
+
+    /* Initialize the clock frequency variable as the CPU's frequency in MHz */
+    sys_clock_freq_mhz = opal_timer_base_get_freq() / 1000000;
+
+    events_init();
+
+    /* Get the MCA params string of counters to turn on */
+    char **arg_strings = opal_argv_split(ompi_mpi_spc_attach_string, ',');
+    int num_args      = opal_argv_count(arg_strings);
+
+    /* If there is only one argument and it is 'all', then all counters
+     * should be turned on.  If the size is 0, then no counters will be enabled.
+     */
+    if(num_args == 1) {
+        if(strcmp(arg_strings[0], "all") == 0)
+            all_on = 1;
+    }
+
+    int prev = -1;
+    /* Turn on only the counters that were specified in the MCA parameter */
+    for(i = 0; i < OMPI_NUM_COUNTERS; i++) {
+        if(all_on)
+            attached_event[i] = 1;
+        else {
+            /* Note: If no arguments were given, this will be skipped */
+            for(j = 0; j < num_args && found < num_args; j++) {
+                if(strcmp(counter_names[i], arg_strings[j]) == 0) {
+                    attached_event[i] = 1;
+                    found++;
```

Comment:
Is this used anywhere?  If so, do you need to also increment it in the `all_on` conditional block?